### DISCUSSION TITLE
Trouble with line endings when patching NVIDIA.common.dsc.inc in edk2-firmware-tegra

### DISCUSSION THREAD

**Discussion by bhagen55** [OE4T/meta-tegra#1303] [Q&A] [answered]
Hello,
I am attempting to disable the 5 second UEFI autoboot delay to save some boot time.
That is done, according to Nvidia FAQ, like this:
```
Index: edk2-firmware-tegra/edk2-nvidia/Platform/NVIDIA/NVIDIA.common.dsc.inc
===================================================================
--- edk2-firmware-tegra.orig/edk2-nvidia/Platform/NVIDIA/NVIDIA.common.dsc.inc
+++ edk2-firmware-tegra/edk2-nvidia/Platform/NVIDIA/NVIDIA.common.dsc.inc
@@ -628,7 +628,7 @@
 
 [PcdsDynamicHii.common.DEFAULT]
   # Global variables
-  gEfiMdePkgTokenSpaceGuid.PcdPlatformBootTimeOut|L"Timeout"|gEfiGlobalVariableGuid|0x0|5
+  gEfiMdePkgTokenSpaceGuid.PcdPlatformBootTimeOut|L"Timeout"|gEfiGlobalVariableGuid|0x0|0
   # NVIDIA internal variables
   gNVIDIATokenSpaceGuid.PcdPcieResourceConfigNeeded|L"PCIeResourceConfig"|gNVIDIATokenSpaceGuid|0x0|0|BS
   gNVIDIATokenSpaceGuid.PcdPcieEntryInAcpiConfigNeeded|L"PCIeInOsConfig"|gNVIDIATokenSpaceGuid|0x0|0|BS
```
I ran into https://github.com/OE4T/meta-tegra/issues/1046, and worked around it by removing the patch files from `meta-tegra/recipes-bsp/uefi/edk2-firmware-tegra-35.3.1.inc` so I could get the source code expanded, then generated diff patches manually rather than using `devtool update-recipe`.

I generated the patch using `quilt`, and am running into line ending issues trying to get it to apply. `NVIDIA.common.dsc.inc` uses CLRF line endings, and the patch that `quilt` generated also has `CLRF` line endings, but I think it might be stripping them prior to applying.

```
patching file edk2-nvidia/Platform/NVIDIA/NVIDIA.common.dsc.inc
Hunk #1 FAILED at 628 (different line endings).
1 out of 1 hunk FAILED -- rejects in file edk2-nvidia/Platform/NVIDIA/NVIDIA.common.dsc.inc
Patch 0001-Set-UEFI-boot-timeout-to-0.patch does not apply (enforce with -f)
```

I have tried swapping around the line endings of the patch file to just LF, but that doesn't seem to change anything.
Do I need to pre/append `do_patch` to force it to do a binary patch somehow, or ignore whitespace? I am struggling a bit with `quilt` as I am more used to the `git` patching process.

I am using `kirkstone` currently.

**Comment by dwalkes** [comment]
> I could get the source code expanded, then generated diff patches manually rather than using devtool update-recipe.

> I generated the patch using quilt

Not sure how to get quilt to play nicely with linefeeds, but in the past I've just used `git format-patch -1` to generate the patch with git on edk2-firmware (possibly with the `--relative` option, I can't remember for sure), then moved that patch into the layer by hand.

**Reply by bhagen55** [reply]
I also tried generating the patch with `git format-patch` and ran into the same problem unfortunately.

**Reply by madisongh** [reply, marked as answer]
The DOS-style line endings in the UEFI code base, the multiple repos involved, and the fact that `${S}` is the top of the `edk2` repo (with the other repos sitting beside, rather than below), all contribute to the patch-generation problem.  `git format-patch` does work best, but you may have to massage the paths a bit.  For consistency, regardless of which of the component repos has the patch, I'll use
```$ git format-patch --src-prefix="edk2-tegra.a/edk2-nvidia/" --dst-prefix="edk2-tegra.b/edk2-nvidia/" ...```
to generate the patches (with that example being for the `edk2-nvidia` repo), and then add `;patchdir=..` to the entry in SRC_URI, so that the patching is applied from the top-level directory above where the source repos get checked out.

**Reply by bhagen55** [reply]
I also ended up editing the file using vim instead of VSCode, as it seemed to be doing some line ending trickery behind the scenes.
